core/shared/src/main/scala/scalaz/zio/IO.scala

```diff
@@ -534,13 +554,16 @@ sealed abstract class IO[+E, +A] extends Serializable { self =>
    * }}}
    */
   final def timeout[B](z: B)(f: A => B)(duration: Duration): IO[E, B] =
-    self
-      .map(f)
-      .attempt
-      .raceWith[E, Either[E, B], B, B](IO.now[B](z).delay(duration))(
-        (either: Either[E, B], right: Fiber[E, B]) => right.interrupt *> either.fold(IO.fail, IO.now),
-        (b: B, left: Fiber[E, Either[E, B]]) => left.interrupt *> IO.now(b)
-      )
+    IO.now(z).delay(duration).fork.flatMap { fiber =>
+      self
+        .map(f)
+        .attempt
+        .onTermination(_ => fiber.interrupt)
```

Comment:
These are our desired semantics:

1. If `self` takes longer than `duration`, then `z` is chosen.
2. In all other cases, `self` is chosen as soon as possible.

I think we can do that with:

```scala
self.map(f).sandboxWith(io => 
  IO.absolve(IO.now(z).delay(duration) race io.attempt))
```

The sandboxing ensures that `self` will "win" the race whether it fails with some value, or terminates due to defect or interruption.

Does that seem correct to you?